File: algorithms/searching/linear_search/java/LinearSearch.java
/*
 * LinearSearch.java
 * Implementation of Linear Search in Java
 * 
 * Description:
 * Linear Search sequentially checks each element in an array to find a target value.
 * Returns the index of the target if found, otherwise returns -1.
 * 
 * Time Complexity: O(n) - worst case, every element is checked once
 * Space Complexity: O(1) - no additional memory used
 */

public class LinearSearch {

    /**
     * Performs linear search on the given array.
     *
     * @param arr    The array to search through
     * @param target The value to search for
     * @return Index of target if found; -1 otherwise
     */
    public static int linearSearch(int[] arr, int target) {
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] == target) {
                return i; // Target found, return its index
            }
        }
        return -1; // Target not found
    }

    /**
     * Example usage of the Linear Search function
     */
    public static void main(String[] args) {
        int[] numbers = {5, 10, 15, 20, 25};
        int target = 15;

        int index = linearSearch(numbers, target);

        if (index != -1) {
            System.out.println("Target " + target + " found at index: " + index);
        } else {
            System.out.println("Target " + target + " not found in the array.");
        }
    }
}

File: algorithms/searching/linear_search/python/linear_search.py
#
# Linear Search Algorithm in Python
#

# Time Complexity: O(n) - In the worst case, we have to scan the entire list.
# Space Complexity: O(1) - No extra space is needed.
#

def linear_search(arr, target):
    """
    Finds the index of a target element in an array using Linear Search.

    Args:
        arr: A list of elements.
        target: The element to search for.

    Returns:
        The index of the target element if found, otherwise -1.
    """
    for i in range(len(arr)):
        if arr[i] == target:
            return i  # Return the index where the target was found
    return -1  # Return -1 if the target is not in the list


# Example Usage:
if __name__ == "__main__":
    sample_array = [10, 50, 30, 70, 80, 20, 90, 40]
    search_target = 20

    result = linear_search(sample_array, search_target)

    if result != -1:
        print(f"Element {search_target} is present at index {result}")
    else:
        print(f"Element {search_target} is not present in the array")